README.md
# GPIO Peripheral

This repository provides an simple GPIO peripheral with integrated register file
to control 64 GPIOs (by default). The peripheral performs two stage
synchronization of the inputs to resolve potential metastability. The outputs
can be driven in push-pull or open-drain mode. Each GPIO supports any
combination of rising-edge, falling-edge, low-level and high-level interrupts
with individual status registers to query the type of pending interrupts.

The interface to the peripheral is the lightweight [register_interface protocol](https://github.com/pulp-platform/register_interface ).
However, the repository contains convenience wrappers to attach AXI-lite or APB
buses for control. Each module in the repository contains an additional wrapper
at the bottom of the respective source files for the users that prefer
SystemVerilog interfaces over hierarchical structs.

# Changing Number of GPIOs
  Changing the number of GPIOs requires regeneration of the register file to
  include the right number of config registers. The `gpio.sv` will automatically
  adapt accordingly.
  
  The repo contains a Makefile that simplifies the process of invoking the reggen tool for this regeneration.
  E.g. the following comand will reconfigure the project for 48 GPIOs.
  
  ```
  make reconfigure GPIOS=48
  ```

# Ports
| **Signal Name**  | **Direction** | **Description**                                                                                                                                               |
| ---------------  | ------------- | ---------------                                                                                                                                               |
| `clk_i`          | *input*       | Primary input clock. The control interface is suposed to be synchronous to this clock.                                                                        |
| `rst_ni`         | *input*       | Asynchronous active-low reset                                                                                                                                 |
| `gpio_in`        | *input*       | GPIO input signals from IO Pads (Pad -> SoC) signal.                                                                                                          |
| `gpio_out`       | *output*      | GPIO output signals to IO Pads (SoC -> Pad) signal.                                                                                                           |
| `gpio_tx_en_o`   | *output*      | GPIO TX Buffer enable signal. This signal is supposed to control the output buffer enable of the IO Pad. 0 -> TX disabled (High-Z or Pull-low/high), 1 -> TX. |
| `gpio_in_sync_o` | *input*       | Synchronized GPIO input signals. This port provides the `gpio_in` signal synchronized to `clk_i`.                                                             |
| `interrupt_o`    | *output*      | Global interrupt line. The interrupt line is asserted for one `clk_i` whenever an unmasked interrupt on one of the GPIOs arrives.                             |
| `reg_req_i`      | *input*       | Control interface request side using register interface protocol.                                                                                             |
| `reg_rsp_o`      | *output*      | Control interface request side using register_interface protocol.                                                                                             |

# Clock Gates
The GPIO IP manually instantiates a clock gate for each input to reduce power
consumption when the corresponding GPIO is disabled. However, some target
technologies (e.g. FPGA) don't behave that well when the clock path contains to
many clock gating resources. Therefore, the IP is available in two flavors, one
with the manual clock gates and one without them. You can (*and have to*) select
between the two by either supplying the /Bender Target/ `-t gpio_with_clk_gates`
or `-t gpio_no_clk_gates`.

# Register Map
The registers of this module are all defined in the `gpio_regs.hjson` file which
is used to auto-generate the actual SV register file using [lowRISCs reggen tool](https://docs.opentitan.org/doc/rm/register_tool/ ).

Here is a summary of the registers:
## `INFO` Register (offset 0x00, read-only)
  Contains read-only registers with the number of GPIOs this instance of the
  GPIO peripheral was parametrized for and an IP version number.
  
  | 31 - 20    | 19 - 10    | 9 - 0           |
  |------------|------------|-----------------|
  | *reserved* | IP_VERSION | Number of GPIOs |


## `CFG` Register (offset 0x04, rw)
  Controls the interrupt mode of the gpios. If 1, keep the interrupt line
  asserted until all interrupts for all GPIOs are cleared. If 0, generate one
  cycle wide pulses for every new interrupt.

  | 31 - 1     | 0                |
  |------------|------------------|
  | *reserved* | `interrupt_mode` |

## `GPIO_MODE<0-XX>`  Registers (offset 0x08+<reg_idx>*4, rw)
  The GPIO_MODE registers control the operating mode of the individual GPIOs.
  Each register controls 16 GPIOs. 

   | 31 - 30  | ... | 3 - 2   | 1 - 0   |
   |----------|-----|---------|---------|
   | `GPIO15` | ... | `GPIO1` | `GPIO0` |
  
   *Continues in next register*

  The values for `GPIO0_MODE` are:
  
  | Value | Description                                             |
  |-------|---------------------------------------------------------|
  | 0     | Configures GPIO as an input.                            |
  | 1     | Configures GPIO as a push-pull output.                  |
  | 2     | Configures the GPIO to be in open_drain (drive 1) mode. |
  | 3     | Configures the GPIO to be in open_drain (drive 0) mode. |

## `GPIO_EN<0-XX>` Registers (offset 0x80+<reg_idx>*4, rw)
  Each bit of these registers control the sampling of one GPIO. This register
  enables sampling of the inputs. If disables (0) the corresponding GPIO will
  not sample the inputs (saves power) and will not generate any interrupts.
  
   |       31 | ... |       1 |       0 |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `GPIO_IN<0-XX>` Registers (offset 0x100+<reg_idx>*4, read-only)
  The bits of these registers contain the input values of the corresponding
  gpios. 

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `GPIO_OUT<0-XX>` Registers (offset 0x180+<reg_idx>*4, rw)
  Change the value of the corresponding GPIO. In `GPIO_MODE 1` writing 1 drives
  high, writing 0 drives low. In `GPIO_MODE 2` (drive 1) writing a 1 will drive
  the gpio to high while writing a 0 will put the gpio in high-z. In `GPIO_MODE
  3` writing 0 drives low and writing 1 puts the gpio into high-z.
  
   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

  *Continues in next register*

## `GPIO_SET<0-XX>` Registers (offset 0x200+<reg_idx>*4, rw)
  For each asserted bit in the register, set the corresponding bit in the
  GPIO_OUT register (masked set). This simplifys setting a single gpio without
  altering the state of the other ones.
  
   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `GPIO_CLEAR<0-XX>` Registers (offset 0x280+<reg_idx>*4, rw)
  For each asserted bit in the register, clear the corresponding bit in the
  GPIO_OUT register (masked clear). This simplifys clearing a single gpio without
  altering the state of the other ones.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `GPIO_TOGGLE<0-XX>` Registers (offset 0x300+<reg_idx>*4, rw)
  For each asserted bit in the register, toggle the corresponding bit in the
  GPIO_OUT register (masked toggle). This simplifys toggling a single gpio without
  altering the state of the other ones.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_RISE_EN<0-xx>` Registers (offset 0x380+<reg_idx>*4, rw)
   Enable interrupts on rising edges for the corresponding GPIO. 
   
   The corresponding gpio needs to be enabled `GPIO_EN`. Once an interrupt
   condition is detected, the global interrupt line `interrupt_o` is asserted
   (according to the interrupt mode in `CFG`) and the corresponding bit in the
   interrupt status registers is set. To clear the interrupt, write a `1` to the
   corresponding bit in the status register.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_FALL_EN<0-xx>` Registers (offset 0x400+<reg_idx>*4, rw)
   Enable interrupts on falling edges for the corresponding GPIO. 
   
   The corresponding gpio needs to be enabled `GPIO_EN`. Once an interrupt
   condition is detected, the global interrupt line `interrupt_o` is asserted
   (according to the interrupt mode in `CFG`) and the corresponding bit in the
   interrupt status registers is set. To clear the interrupt, write a `1` to the
   corresponding bit in the status register.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_LVL_HIGH_EN<0-xx>` Registers (offset 0x480+<reg_idx>*4, rw)
   Enable logic-high level-sensitive interrupts for the corresponding GPIO
   (interrupt keeps getting triggered while GPIO is high).
   
   The corresponding gpio needs to be enabled `GPIO_EN`. Once an interrupt
   condition is detected, the global interrupt line `interrupt_o` is asserted
   (according to the interrupt mode in `CFG`) and the corresponding bit in the
   interrupt status registers is set. To clear the interrupt, write a `1` to the
   corresponding bit in the status register.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_LVL_LOW_EN<0-xx>` Registers (offset 0x500+<reg_idx>*4, rw)
   Enable logic-low level-sensitive interrupts for the corresponding GPIO
   (interrupt keeps getting triggered while GPIO is low).
   
   The corresponding gpio needs to be enabled `GPIO_EN`. Once an interrupt
   condition is detected, the global interrupt line `interrupt_o` is asserted
   (according to the interrupt mode in `CFG`) and the corresponding bit in the
   interrupt status registers is set. To clear the interrupt, write a `1` to the
   corresponding bit in the status register.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_STATUS<0-xx>` Registers (offset 0x580+<reg_idx>*4, rw)
   Each bit indicates if there are any pending interrupts on the corresponding
   GPIO. Writing a 1 to a specific bit clears **all** pending interrupts (rise,
   fall, low, high) for the corresponding GPIO.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_RISE_STATUS<0-xx>` Registers (offset 0x600+<reg_idx>*4, rw)
   Each bit indicates if there is a pending rising-edge interrupt on the corresponding
   GPIO. Writing a 1 to a specific bit clears the interrupt for the corresponding GPIO.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_FALL_STATUS<0-xx>` Registers (offset 0x680+<reg_idx>*4, rw)
   Each bit indicates if there is a pending falling-edge interrupt on the corresponding
   GPIO. Writing a 1 to a specific bit clears the interrupt for the corresponding GPIO.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_LVL_LOW_STATUS<0-xx>` Registers (offset 0x700+<reg_idx>*4, rw)
   Each bit indicates if there is a pending low-level sensitive interrupt on the corresponding
   GPIO. Writing a 1 to a specific bit clears the interrupt for the corresponding GPIO.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*

## `INTRPT_LVL_HIGH_STATUS<0-xx>` Registers (offset 0x780+<reg_idx>*4, rw)
   Each bit indicates if there is a pending low-level sensitive interrupt on the corresponding
   GPIO. Writing a 1 to a specific bit clears the interrupt for the corresponding GPIO.

   | 31       | ... | 1       | 0       |
   |----------|-----|---------|---------|
   | `GPIO31` | ... | `GPIO1` | `GPIO0` |

   *Continues in next register*



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
### Changed
- Bump bender dependencies
### Fixed
- Fix typos and errors in documentation

## 0.2.1 - 2022-12-15
### Changed
- Simplified reconfiguration flow by using vendored-in regtool.py
### Fixed
- Overlapping bitfield definition in hjson description of the CFG register. 
- Remove now obsolete Makefile dependency on bender checkout dir

## 0.2.0 - 2022-12-09
### Breaking Changes
- Changed the module interface. ``interrupt_o`` was renamed to
  ``global_interrupt_o``. Additionally, the module also exposes pin level
  interrupt signals in addition to the single, globally multiplexed gpio
  interrupt signal. Existing RTL integrating this IP need to adapt the port
  list of their instantiations.
### Changed
- Change default pad count from 56 to 32.
- Use the clock gated input stage by default for simulation targets
- Bump AXI Version
### Fixed
- Fix warning about unconnected interface port

## 0.1.2 - 2022-12-04
### Changed
- Added make dependencies to auto-setup python env for reconfiguration

### Fixed
- Fix some small issues reported by linter

## 0.1.1 - 2022-10-07
### Changed
- Bumped AXI version to v0.35.3
- Added NumRepetitions to tb_gpio to choose test duration
- Refactored TB

### Fixed
- Fix tx_en inversion bug for open-drain mode 1
- Fix bug in TB that caused open-drain misbehavior not to be catched


## 0.1.0 - 2022-04-14
Initial release


src/gpio_reg_pkg.sv
// Copyright lowRISC contributors.
// Licensed under the Apache License, Version 2.0, see LICENSE for details.
// SPDX-License-Identifier: Apache-2.0
//
// Register Package auto-generated by `reggen` containing data structure

package gpio_reg_pkg;

  // Param list
  parameter int GPIOCount = 56;

  // Address widths within the block
  parameter int BlockAw = 11;

  ////////////////////////////
  // Typedefs for registers //
  ////////////////////////////

  typedef struct packed {
    struct packed {
      logic        q;
    } intrpt_mode;
    struct packed {
      logic        q;
    } reserved;
  } gpio_reg2hw_cfg_reg_t;

  typedef struct packed {
    logic [1:0]  q;
  } gpio_reg2hw_gpio_mode_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_gpio_en_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_gpio_out_mreg_t;

  typedef struct packed {
    logic        q;
    logic        qe;
  } gpio_reg2hw_gpio_set_mreg_t;

  typedef struct packed {
    logic        q;
    logic        qe;
  } gpio_reg2hw_gpio_clear_mreg_t;

  typedef struct packed {
    logic        q;
    logic        qe;
  } gpio_reg2hw_gpio_toggle_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_rise_en_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_fall_en_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_lvl_high_en_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_lvl_low_en_mreg_t;

  typedef struct packed {
    logic        q;
    logic        qe;
  } gpio_reg2hw_intrpt_status_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_rise_status_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_fall_status_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_lvl_high_status_mreg_t;

  typedef struct packed {
    logic        q;
  } gpio_reg2hw_intrpt_lvl_low_status_mreg_t;

  typedef struct packed {
    struct packed {
      logic [9:0] d;
    } gpio_cnt;
    struct packed {
      logic [9:0] d;
    } version;
  } gpio_hw2reg_info_reg_t;

  typedef struct packed {
    logic        d;
  } gpio_hw2reg_gpio_in_mreg_t;

  typedef struct packed {
    logic        d;
    logic        de;
  } gpio_hw2reg_gpio_out_mreg_t;

  typedef struct packed {
    logic        d;
  } gpio_hw2reg_intrpt_status_mreg_t;

  typedef struct packed {
    logic        d;
    logic        de;
  } gpio_hw2reg_intrpt_rise_status_mreg_t;

  typedef struct packed {
    logic        d;
    logic        de;
  } gpio_hw2reg_intrpt_fall_status_mreg_t;

  typedef struct packed {
    logic        d;
    logic        de;
  } gpio_hw2reg_intrpt_lvl_high_status_mreg_t;

  typedef struct packed {
    logic        d;
    logic        de;
  } gpio_hw2reg_intrpt_lvl_low_status_mreg_t;

  // Register -> HW type
  typedef struct packed {
    gpio_reg2hw_cfg_reg_t cfg; // [1121:1120]
    gpio_reg2hw_gpio_mode_mreg_t [55:0] gpio_mode; // [1119:1008]
    gpio_reg2hw_gpio_en_mreg_t [55:0] gpio_en; // [1007:952]
    gpio_reg2hw_gpio_out_mreg_t [55:0] gpio_out; // [951:896]
    gpio_reg2hw_gpio_set_mreg_t [55:0] gpio_set; // [895:784]
    gpio_reg2hw_gpio_clear_mreg_t [55:0] gpio_clear; // [783:672]
    gpio_reg2hw_gpio_toggle_mreg_t [55:0] gpio_toggle; // [671:560]
    gpio_reg2hw_intrpt_rise_en_mreg_t [55:0] intrpt_rise_en; // [559:504]
    gpio_reg2hw_intrpt_fall_en_mreg_t [55:0] intrpt_fall_en; // [503:448]
    gpio_reg2hw_intrpt_lvl_high_en_mreg_t [55:0] intrpt_lvl_high_en; // [447:392]
    gpio_reg2hw_intrpt_lvl_low_en_mreg_t [55:0] intrpt_lvl_low_en; // [391:336]
    gpio_reg2hw_intrpt_status_mreg_t [55:0] intrpt_status; // [335:224]
    gpio_reg2hw_intrpt_rise_status_mreg_t [55:0] intrpt_rise_status; // [223:168]
    gpio_reg2hw_intrpt_fall_status_mreg_t [55:0] intrpt_fall_status; // [167:112]
    gpio_reg2hw_intrpt_lvl_high_status_mreg_t [55:0] intrpt_lvl_high_status; // [111:56]
    gpio_reg2hw_intrpt_lvl_low_status_mreg_t [55:0] intrpt_lvl_low_status; // [55:0]
  } gpio_reg2hw_t;

  // HW -> register type
  typedef struct packed {
    gpio_hw2reg_info_reg_t info; // [691:672]
    gpio_hw2reg_gpio_in_mreg_t [55:0] gpio_in; // [671:616]
    gpio_hw2reg_gpio_out_mreg_t [55:0] gpio_out; // [615:504]
    gpio_hw2reg_intrpt_status_mreg_t [55:0] intrpt_status; // [503:448]
    gpio_hw2reg_intrpt_rise_status_mreg_t [55:0] intrpt_rise_status; // [447:336]
    gpio_hw2reg_intrpt_fall_status_mreg_t [55:0] intrpt_fall_status; // [335:224]
    gpio_hw2reg_intrpt_lvl_high_status_mreg_t [55:0] intrpt_lvl_high_status; // [223:112]
    gpio_hw2reg_intrpt_lvl_low_status_mreg_t [55:0] intrpt_lvl_low_status; // [111:0]
  } gpio_hw2reg_t;

  // Register offsets
  parameter logic [BlockAw-1:0] GPIO_INFO_OFFSET = 11'h 0;
  parameter logic [BlockAw-1:0] GPIO_CFG_OFFSET = 11'h 4;
  parameter logic [BlockAw-1:0] GPIO_GPIO_MODE_0_OFFSET = 11'h 8;
  parameter logic [BlockAw-1:0] GPIO_GPIO_MODE_1_OFFSET = 11'h c;
  parameter logic [BlockAw-1:0] GPIO_GPIO_MODE_2_OFFSET = 11'h 10;
  parameter logic [BlockAw-1:0] GPIO_GPIO_MODE_3_OFFSET = 11'h 14;
  parameter logic [BlockAw-1:0] GPIO_GPIO_EN_0_OFFSET = 11'h 80;
  parameter logic [BlockAw-1:0] GPIO_GPIO_EN_1_OFFSET = 11'h 84;
  parameter logic [BlockAw-1:0] GPIO_GPIO_IN_0_OFFSET = 11'h 100;
  parameter logic [BlockAw-1:0] GPIO_GPIO_IN_1_OFFSET = 11'h 104;
  parameter logic [BlockAw-1:0] GPIO_GPIO_OUT_0_OFFSET = 11'h 180;
  parameter logic [BlockAw-1:0] GPIO_GPIO_OUT_1_OFFSET = 11'h 184;
  parameter logic [BlockAw-1:0] GPIO_GPIO_SET_0_OFFSET = 11'h 200;
  parameter logic [BlockAw-1:0] GPIO_GPIO_SET_1_OFFSET = 11'h 204;
  parameter logic [BlockAw-1:0] GPIO_GPIO_CLEAR_0_OFFSET = 11'h 280;
  parameter logic [BlockAw-1:0] GPIO_GPIO_CLEAR_1_OFFSET = 11'h 284;
  parameter logic [BlockAw-1:0] GPIO_GPIO_TOGGLE_0_OFFSET = 11'h 300;
  parameter logic [BlockAw-1:0] GPIO_GPIO_TOGGLE_1_OFFSET = 11'h 304;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_RISE_EN_0_OFFSET = 11'h 380;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_RISE_EN_1_OFFSET = 11'h 384;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_FALL_EN_0_OFFSET = 11'h 400;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_FALL_EN_1_OFFSET = 11'h 404;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_HIGH_EN_0_OFFSET = 11'h 480;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_HIGH_EN_1_OFFSET = 11'h 484;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_LOW_EN_0_OFFSET = 11'h 500;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_LOW_EN_1_OFFSET = 11'h 504;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_STATUS_0_OFFSET = 11'h 580;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_STATUS_1_OFFSET = 11'h 584;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_RISE_STATUS_0_OFFSET = 11'h 600;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_RISE_STATUS_1_OFFSET = 11'h 604;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_FALL_STATUS_0_OFFSET = 11'h 680;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_FALL_STATUS_1_OFFSET = 11'h 684;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_HIGH_STATUS_0_OFFSET = 11'h 700;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_HIGH_STATUS_1_OFFSET = 11'h 704;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_LOW_STATUS_0_OFFSET = 11'h 780;
  parameter logic [BlockAw-1:0] GPIO_INTRPT_LVL_LOW_STATUS_1_OFFSET = 11'h 784;

  // Reset values for hwext registers and their fields
  parameter logic [19:0] GPIO_INFO_RESVAL = 20'h 800;
  parameter logic [9:0] GPIO_INFO_VERSION_RESVAL = 10'h 2;
  parameter logic [31:0] GPIO_GPIO_IN_0_RESVAL = 32'h 0;
  parameter logic [23:0] GPIO_GPIO_IN_1_RESVAL = 24'h 0;
  parameter logic [31:0] GPIO_GPIO_SET_0_RESVAL = 32'h 0;
  parameter logic [23:0] GPIO_GPIO_SET_1_RESVAL = 24'h 0;
  parameter logic [31:0] GPIO_GPIO_CLEAR_0_RESVAL = 32'h 0;
  parameter logic [23:0] GPIO_GPIO_CLEAR_1_RESVAL = 24'h 0;
  parameter logic [31:0] GPIO_GPIO_TOGGLE_0_RESVAL = 32'h 0;
  parameter logic [23:0] GPIO_GPIO_TOGGLE_1_RESVAL = 24'h 0;
  parameter logic [31:0] GPIO_INTRPT_STATUS_0_RESVAL = 32'h 0;
  parameter logic [23:0] GPIO_INTRPT_STATUS_1_RESVAL = 24'h 0;

  // Register index
  typedef enum int {
    GPIO_INFO,
    GPIO_CFG,
    GPIO_GPIO_MODE_0,
    GPIO_GPIO_MODE_1,
    GPIO_GPIO_MODE_2,
    GPIO_GPIO_MODE_3,
    GPIO_GPIO_EN_0,
    GPIO_GPIO_EN_1,
    GPIO_GPIO_IN_0,
    GPIO_GPIO_IN_1,
    GPIO_GPIO_OUT_0,
    GPIO_GPIO_OUT_1,
    GPIO_GPIO_SET_0,
    GPIO_GPIO_SET_1,
    GPIO_GPIO_CLEAR_0,
    GPIO_GPIO_CLEAR_1,
    GPIO_GPIO_TOGGLE_0,
    GPIO_GPIO_TOGGLE_1,
    GPIO_INTRPT_RISE_EN_0,
    GPIO_INTRPT_RISE_EN_1,
    GPIO_INTRPT_FALL_EN_0,
    GPIO_INTRPT_FALL_EN_1,
    GPIO_INTRPT_LVL_HIGH_EN_0,
    GPIO_INTRPT_LVL_HIGH_EN_1,
    GPIO_INTRPT_LVL_LOW_EN_0,
    GPIO_INTRPT_LVL_LOW_EN_1,
    GPIO_INTRPT_STATUS_0,
    GPIO_INTRPT_STATUS_1,
    GPIO_INTRPT_RISE_STATUS_0,
    GPIO_INTRPT_RISE_STATUS_1,
    GPIO_INTRPT_FALL_STATUS_0,
    GPIO_INTRPT_FALL_STATUS_1,
    GPIO_INTRPT_LVL_HIGH_STATUS_0,
    GPIO_INTRPT_LVL_HIGH_STATUS_1,
    GPIO_INTRPT_LVL_LOW_STATUS_0,
    GPIO_INTRPT_LVL_LOW_STATUS_1
  } gpio_id_e;

  // Register width information to check illegal writes
  parameter logic [3:0] GPIO_PERMIT [36] = '{
    4'b 0111, // index[ 0] GPIO_INFO
    4'b 0001, // index[ 1] GPIO_CFG
    4'b 1111, // index[ 2] GPIO_GPIO_MODE_0
    4'b 1111, // index[ 3] GPIO_GPIO_MODE_1
    4'b 1111, // index[ 4] GPIO_GPIO_MODE_2
    4'b 0011, // index[ 5] GPIO_GPIO_MODE_3
    4'b 1111, // index[ 6] GPIO_GPIO_EN_0
    4'b 0111, // index[ 7] GPIO_GPIO_EN_1
    4'b 1111, // index[ 8] GPIO_GPIO_IN_0
    4'b 0111, // index[ 9] GPIO_GPIO_IN_1
    4'b 1111, // index[10] GPIO_GPIO_OUT_0
    4'b 0111, // index[11] GPIO_GPIO_OUT_1
    4'b 1111, // index[12] GPIO_GPIO_SET_0
    4'b 0111, // index[13] GPIO_GPIO_SET_1
    4'b 1111, // index[14] GPIO_GPIO_CLEAR_0
    4'b 0111, // index[15] GPIO_GPIO_CLEAR_1
    4'b 1111, // index[16] GPIO_GPIO_TOGGLE_0
    4'b 0111, // index[17] GPIO_GPIO_TOGGLE_1
    4'b 1111, // index[18] GPIO_INTRPT_RISE_EN_0
    4'b 0111, // index[19] GPIO_INTRPT_RISE_EN_1
    4'b 1111, // index[20] GPIO_INTRPT_FALL_EN_0
    4'b 0111, // index[21] GPIO_INTRPT_FALL_EN_1
    4'b 1111, // index[22] GPIO_INTRPT_LVL_HIGH_EN_0
    4'b 0111, // index[23] GPIO_INTRPT_LVL_HIGH_EN_1
    4'b 1111, // index[24] GPIO_INTRPT_LVL_LOW_EN_0
    4'b 0111, // index[25] GPIO_INTRPT_LVL_LOW_EN_1
    4'b 1111, // index[26] GPIO_INTRPT_STATUS_0
    4'b 0111, // index[27] GPIO_INTRPT_STATUS_1
    4'b 1111, // index[28] GPIO_INTRPT_RISE_STATUS_0
    4'b 0111, // index[29] GPIO_INTRPT_RISE_STATUS_1
    4'b 1111, // index[30] GPIO_INTRPT_FALL_STATUS_0
    4'b 0111, // index[31] GPIO_INTRPT_FALL_STATUS_1
    4'b 1111, // index[32] GPIO_INTRPT_LVL_HIGH_STATUS_0
    4'b 0111, // index[33] GPIO_INTRPT_LVL_HIGH_STATUS_1
    4'b 1111, // index[34] GPIO_INTRPT_LVL_LOW_STATUS_0
    4'b 0111  // index[35] GPIO_INTRPT_LVL_LOW_STATUS_1
  };

endpackage

